### PR TITLE
[several] Various small improvements (main)

### PR DESCRIPTION
**Pull request by SwooshyCueb**
Addresses #5803
Addresses #6237
Addresses #6528
Addresses #7508
Addresses #7764
Addresses #7766
Addresses #7767
Addresses #7768
Addresses #7769
Addresses #7781
Addresses #7800
Addresses #7802
Addresses #7803
Addresses #7809
Addresses #7812

### CODE REVIEW

**Review by alanking — COMMENTED** [review]
Looks good overall

File: lib/core/include/irods/irods_logger.hpp [hunk 1]
```diff
@@ -656,29 +663,34 @@ namespace irods::experimental::log
...
-                // clang-format on
-
-                timespec ts; // NOLINT(cppcoreguidelines-pro-type-member-init)
-
-                if (const auto ec = clock_gettime(CLOCK_REALTIME, &ts); ec != 0) {
-                    const auto now = clock_type::to_time_t(clock_type::now());
-                    std::stringstream utc_ss;
-                    utc_ss << std::put_time(std::gmtime(&now), "%FT%T.000Z"); // NOLINT(concurrency-mt-unsafe)
-                    return utc_ss.str();
+                namespace chrono = std::chrono;
```
[the remark this answers is not in the document]
Comment (SwooshyCueb) [reply lib/core/include/irods/irods_logger.hpp:666]: Yes

File: lib/core/include/irods/irods_logger.hpp [hunk 2]
```diff
@@ -656,29 +663,34 @@ namespace irods::experimental::log
...
+                using clock_type = chrono::system_clock;
+
+#if IRODS_CHRONO_FORMATTER_FMT
+                const auto tp = chrono::floor<chrono::milliseconds>(clock_type::now());
+                return fmt::format(FMT_COMPILE("{:%FT%H:%M:%S}Z"), tp);
+#else
+                char ts_buf[64]{};
```
[the remark this answers is not in the document]
Comment (SwooshyCueb) [reply lib/core/include/irods/irods_logger.hpp:673]: Smallest power of two I feel comfortable using for the buffer size.

File: unit_tests/CMakeLists.txt
[the remark this answers is not in the document]
Comment (SwooshyCueb) [reply unit_tests/CMakeLists.txt:1]: I don't mind changing it

File: lib/core/include/irods/irods_logger.hpp [hunk 3]
```diff
@@ -484,14 +485,49 @@ namespace irods::experimental::log
             ///
             /// \since 4.3.0
             template <typename... Args>
-            auto operator()(fmt::format_string<Args...> _format, Args&&... _args) const -> void
+            void operator()(fmt::format_string<Args...> _format, Args&&... _args) const
```
Comment (korydraughn) [inline lib/core/include/irods/irods_logger.hpp:495]: Let's use the trailing return type form for consistency.

File: lib/core/include/irods/irods_logger.hpp [hunk 4]
```diff
@@ -484,14 +485,49 @@ namespace irods::experimental::log
...
+                // Clang 15 and later (and GCC 11 and later) consider this overload ambiguous
+                // with the std::invocable overload. This constraint sorts that out.
```
Comment (korydraughn) [inline lib/core/include/irods/irods_logger.hpp:497]: Let's move this inside the preprocessor conditional just like the one that follows.

File: lib/core/include/irods/irods_logger.hpp [hunk 5]
```diff
@@ -484,14 +485,49 @@ namespace irods::experimental::log
...
+            // This overload covers the 2-argument cases
+            template <typename Arg>
+            void operator()(fmt::format_string<Arg> _format, Arg&& _arg) const
+                requires (!std::is_invocable_v<Arg>)
+            {
+                if (should_log()) {
+                    const auto msg = {key_value{tag::log::message, fmt::format(_format, std::forward<Args>(_arg))}};
+                    log_message(std::begin(msg), std::end(msg));
+                }
+            }
```
Comment (korydraughn) [inline lib/core/include/irods/irods_logger.hpp:522]: Let's add the `// operator()` to the ends of these functions for consistency.

File: lib/core/include/irods/irods_logger.hpp [hunk 6]
```diff
@@ -656,29 +663,34 @@ namespace irods::experimental::log
...
+                return fmt::format(FMT_COMPILE("{:%FT%H:%M:%S}Z"), tp);
+#else
+                char ts_buf[64]{};
+                std::timespec ts; // NOLINT(cppcoreguidelines-pro-type-member-init)
+                if (const int base = std::timespec_get(&ts, TIME_UTC); base != TIME_UTC) {
+                    if (const int ec = clock_gettime(CLOCK_REALTIME, &ts); ec != 0) {
+                        const auto tp = clock_type::now();
+                        const auto tp_s = chrono::floor<chrono::seconds>(tp);
+                        const auto tp_s_ms = chrono::floor<chrono::milliseconds>(tp_s);
+                        const auto tp_ms = chrono::floor<chrono::milliseconds>(tp) - tp_s_ms;
```
Comment (korydraughn) [inline lib/core/include/irods/irods_logger.hpp:680]: How did we determine `floor` was needed?

**Review by korydraughn — COMMENTED** [review]
Good stuff.

File: lib/core/src/getRodsEnv.cpp [hunk 7]
```diff
@@ -142,13 +142,14 @@ extern "C" {
         createRodsEnvDefaults( &rodsEnvArg );
     }
 
-    static int capture_string_property(const std::string& _key, char* _val, std::size_t _val_size)
+    int capture_string_property(const char* _key, char* _val, size_t _val_size)
     {
+        std::string key{_key};
```
Comment (korydraughn) [inline lib/core/src/getRodsEnv.cpp:147]: `_key` needs to be checked for `nullptr` before constructing the `std::string`.

File: lib/core/include/irods/irods_logger.hpp [hunk 8]
```diff
@@ -484,14 +485,49 @@ namespace irods::experimental::log
...
+                requires (sizeof...(Args) == 0)
+#endif
             {
                 if (should_log()) {
                     const auto msg = {key_value{tag::log::message, fmt::format(_format, std::forward<Args>(_args)...)}};
                     log_message(std::begin(msg), std::end(msg));
                 }
             } // operator()
 
+#if __clang_major__ == 15
```
[the remark this answers is not in the document]
Comment (alanking) [reply lib/core/include/irods/irods_logger.hpp:512]: Makes sense.

File: unit_tests/CMakeLists.txt
[the remark this answers is not in the document]
Comment (alanking) [reply unit_tests/CMakeLists.txt:1]: Okay, let's update that commit message, then. Thanks

File: lib/core/include/irods/irods_logger.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (alanking) [reply lib/core/include/irods/irods_logger.hpp:666]: Understood.

File: lib/core/include/irods/irods_logger.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (alanking) [reply lib/core/include/irods/irods_logger.hpp:673]: Okay. Consider leaving a comment or capturing in a well-named constant so we aren't left wondering if the 64 has some special significance in the future.

File: lib/core/include/irods/irods_logger.hpp (hunk 6 above)
Comment (SwooshyCueb) [reply lib/core/include/irods/irods_logger.hpp:680]: It was either that or `time_point_cast`; they both do effectively the same thing here. I did some timing tests and found that `floor` is marginally faster, at least on my machine.

File: lib/core/include/irods/irods_logger.hpp (hunk 3 above)
Comment (SwooshyCueb) [reply lib/core/include/irods/irods_logger.hpp:495]: I'm not sure how that is done with a constraint.

File: lib/core/include/irods/irods_logger.hpp (hunk 6 above)
Comment (korydraughn) [reply lib/core/include/irods/irods_logger.hpp:680]: Cool cool.

File: lib/core/include/irods/irods_logger.hpp (hunk 3 above)
Comment (korydraughn) [reply lib/core/include/irods/irods_logger.hpp:495]: Ohhh interesting. I experienced a similar thing when attempting to use `override`.

Will look into that. cppreference must state something about it. Perhaps the wording is difficult to find.

File: unit_tests/CMakeLists.txt
[the remark this answers is not in the document]
Comment (SwooshyCueb) [reply unit_tests/CMakeLists.txt:1]: Done

File: lib/core/include/irods/irods_logger.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (SwooshyCueb) [reply lib/core/include/irods/irods_logger.hpp]: Added comment

File: lib/core/include/irods/irods_logger.hpp (hunk 4 above)
Comment (SwooshyCueb) [reply lib/core/include/irods/irods_logger.hpp:498]: Done

File: lib/core/include/irods/irods_logger.hpp (hunk 5 above)
Comment (SwooshyCueb) [reply lib/core/include/irods/irods_logger.hpp]: Done

File: lib/core/src/getRodsEnv.cpp (hunk 7 above)
Comment (SwooshyCueb) [reply lib/core/src/getRodsEnv.cpp]: Done

File: lib/core/include/irods/irods_logger.hpp [hunk 9]
```diff
@@ -484,14 +492,49 @@ namespace irods::experimental::log
...
+                if (should_log()) {
+                    const auto msg = {key_value{tag::log::message, fmt::format(_format, std::forward<Args>(_arg))}};
+                    log_message(std::begin(msg), std::end(msg));
+                }
+            } // operator()
+
+            // This overload covers the 3+-argument cases
+            template <typename Arg1, typename... Args>
+            void operator()(fmt::format_string<Arg1, Args...> _format, Arg1&& _arg1, Args&&... _args) const
+                requires (sizeof...(Args) > 0) // to elinimate ambiguity with previous overload
```
Comment (alanking) [inline lib/core/include/irods/irods_logger.hpp:527]:
```suggestion
                requires (sizeof...(Args) > 0) // to eliminate ambiguity with previous overload
```

**Review by alanking — COMMENTED** [review]
Saw one more thing and otherwise I'm happy with it if you're happy with it. Have you run tests against these changes?

File: lib/core/include/irods/irods_logger.hpp (hunk 9 above)
Comment (SwooshyCueb) [reply lib/core/include/irods/irods_logger.hpp]: Done

**Comment by SwooshyCueb** [comment]
> Have you run tests against these changes?

Am doing so currently

File: lib/core/include/irods/irods_logger.hpp (hunk 3 above)
Comment (SwooshyCueb) [reply lib/core/include/irods/irods_logger.hpp]: I figured it out.

**Comment by SwooshyCueb** [comment]
Core tests pass

File: lib/core/include/irods/irods_logger.hpp [hunk 10]
```diff
@@ -485,13 +493,48 @@ namespace irods::experimental::log
             /// \since 4.3.0
             template <typename... Args>
             auto operator()(fmt::format_string<Args...> _format, Args&&... _args) const -> void
+#if __clang__ == 0 || __clang_major__ > 15
+                // Clang 15 and later (and GCC 11 and later) consider this overload ambiguous
+                // with the std::invocable overload. This constraint sorts that out.
+                requires (sizeof...(Args) != 1 || !std::is_invocable_v<std::tuple_element_t<0, std::tuple<Args...>>>)
+#elif __clang_major__ == 15
+                // Clang 15 can't parse the above statement properly, so we have special overloads for 15.
+                // This overload handles the 1-arguemnt case.
```
Comment (alanking) [inline lib/core/include/irods/irods_logger.hpp:502]: Sorry, I just saw this... Feel free to ignore. :p
```suggestion
                // This overload handles the 1-argument case.
```

**Comment by korydraughn** [comment]
Cool cool. Please run clang-format and we'll take one more pass over it before approving.

**Review by alanking — COMMENTED** [review]
Sweet.

File: lib/core/include/irods/irods_logger.hpp (hunk 10 above)
Comment (SwooshyCueb) [reply lib/core/include/irods/irods_logger.hpp]: Fixed

**Comment by SwooshyCueb** [comment]
> Please run clang-format and we'll take one more pass over it before approving.

Done. The remaining changes suggested by clang-format are either changes I don't care to make, or changes to code not touched by this PR.

**Review by alanking — APPROVED** [review]
Excellent. Let's go ahead and # it

**Comment by SwooshyCueb** [comment]
\#'d

**Pull request merged**